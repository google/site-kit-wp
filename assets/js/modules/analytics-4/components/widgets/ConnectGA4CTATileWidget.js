/**
 * ConnectGA4CTATileWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import ConnectModuleCTATile from '../../../../components/KeyMetrics/ConnectModuleCTATile';
import useWidgetStateEffect from '../../../../googlesitekit/widgets/hooks/useWidgetStateEffect';
import {
	CORE_USER,
	keyMetricsGA4Widgets,
} from '../../../../googlesitekit/datastore/user/constants';
import Null from '../../../../components/Null';

const metadata = { moduleSlug: 'analytics-4' };

export default function ConnectGA4CTATileWidget( { Widget, widgetSlug } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();

		if ( ! keyMetrics ) {
			return [];
		}

		return keyMetrics.filter( ( keyMetric ) =>
			keyMetricsGA4Widgets.includes( keyMetric )
		).length;
	} );

	// Only render the CTA if there are 3 or fewer GA4-dependent key metrics tiles.
	// Otherwise, render `Null` to hide the CTA.
	// If there are four GA4-dependent key metrics tiles, the `ConnectGA4CTAWidget` will be rendered instead.
	const ConnectModuleCTAComponent =
		ga4DependantKeyMetrics > 3 ? Null : ConnectModuleCTATile;

	useWidgetStateEffect( widgetSlug, ConnectModuleCTAComponent, metadata );

	// Note that we need to render the `Widget` component as a wrapper here so this component will display
	// correctly when used as a `FallbackComponent` for the `whenActive` HOC. Conversely, when `ConnectModuleCTATile`
	// is rendered as an `OverrideComponent` in `WidgetRenderer` (as a result of multiple adjacent widgets rendering
	// this component and thus sharing the same state), it is wrapped with a `Widget` component - the net
	// result being the ConnectModuleCTATile is correctly wrapped in a `Widget` in both cases.
	return (
		<Widget>
			<ConnectModuleCTATile { ...metadata } />
		</Widget>
	);
}

ConnectGA4CTATileWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	widgetSlug: PropTypes.string.isRequired,
};
