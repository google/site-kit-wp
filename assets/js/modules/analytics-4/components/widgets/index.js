/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
