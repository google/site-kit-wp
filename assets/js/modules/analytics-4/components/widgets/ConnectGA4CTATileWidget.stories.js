/**
 * ConnectGA4CTATileWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import {
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	provideUserCapabilities,
} from '../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectAdSenseCTATile'
)( ConnectGA4CTATileWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTATileWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTATileWidget',
	delay: 250,
};

export default {
	title: 'Key Metrics/ConnectGA4CTATileWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				provideUserAuthentication( registry );
				provideUserCapabilities( registry );
				provideModules( registry, [
					{
						slug: 'analytics-4',
						active: false,
						connected: false,
					},
				] );
				provideModuleRegistrations( registry );
				provideSiteInfo( registry, {
					postTypes: [ { slug: 'product', label: 'Product' } ],
				} );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
