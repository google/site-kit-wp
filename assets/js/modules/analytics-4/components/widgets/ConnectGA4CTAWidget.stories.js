/**
 * ConnectGA4CTAWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
	KM_ANALYTICS_RETURNING_VISITORS,
	KM_ANALYTICS_NEW_VISITORS,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import { provideKeyMetrics } from '../../../../../../tests/js/utils';
import { provideKeyMetricsWidgetRegistrations } from '../../../../components/KeyMetrics/test-utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import ConnectGA4CTAWidget from './ConnectGA4CTAWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectGA4CTA'
)( ConnectGA4CTAWidget );

function Template() {
	return (
		<div className="googlesitekit-widget-area--mainDashboardKeyMetricsPrimary">
			<div className="googlesitekit-widget--keyMetricsConnectGA4All">
				<WidgetWithComponentProps />
			</div>
		</div>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTAWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTAWidget',
};

export default {
	title: 'Key Metrics/ConnectGA4CTAWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				const keyMetricWidgets = [
					KM_ANALYTICS_RETURNING_VISITORS,
					KM_ANALYTICS_NEW_VISITORS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				];

				provideKeyMetrics( registry, {
					widgetSlugs: keyMetricWidgets,
				} );

				provideKeyMetricsWidgetRegistrations(
					registry,
					keyMetricWidgets.reduce(
						( acc, widget ) => ( {
							...acc,
							[ widget ]: { modules: [ 'analytics-4' ] },
						} ),
						{}
					)
				);
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
