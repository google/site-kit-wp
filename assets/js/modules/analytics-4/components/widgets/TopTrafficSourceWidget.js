/**
 * TopTrafficSourceWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import MetricTileText from '../../../../components/KeyMetrics/MetricTileText';
import {
	CORE_USER,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { numFmt } from '../../../../util';
import { get } from 'lodash';
import whenActive from '../../../../util/when-active';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';

function TopTrafficSourceWidget( { Widget } ) {
	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} )
	);

	const totalUsersReportOptions = {
		...dates,
		metrics: [
			{
				name: 'totalUsers',
			},
		],
	};

	const trafficSourceReportOptions = {
		...dates,
		dimensions: [ 'sessionDefaultChannelGroup' ],
		metrics: [
			{
				name: 'totalUsers',
			},
		],
		limit: 1,
		orderBy: 'totalUsers',
	};

	const totalUsersReport = useInViewSelect(
		( select ) =>
			select( MODULES_ANALYTICS_4 ).getReport( totalUsersReportOptions ),
		[ totalUsersReportOptions ]
	);

	const trafficSourceReport = useInViewSelect(
		( select ) =>
			select( MODULES_ANALYTICS_4 ).getReport(
				trafficSourceReportOptions
			),
		[ trafficSourceReportOptions ]
	);

	const error = useSelect( ( select ) => {
		const trafficSourceReportErrors = select(
			MODULES_ANALYTICS_4
		).getErrorForSelector( 'getReport', [ trafficSourceReportOptions ] );

		const totalUsersReportErrors = select(
			MODULES_ANALYTICS_4
		).getErrorForSelector( 'getReport', [ totalUsersReportOptions ] );

		if ( trafficSourceReportErrors && totalUsersReportErrors ) {
			return [ trafficSourceReportErrors, totalUsersReportErrors ];
		}

		return trafficSourceReportErrors || totalUsersReportErrors || undefined;
	} );

	const loading = useSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ totalUsersReportOptions ]
			) ||
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ trafficSourceReportOptions ]
			)
	);

	const makeFilter = ( dateRange, dimensionIndex ) => ( row ) =>
		get( row, `dimensionValues.${ dimensionIndex }.value` ) === dateRange;

	// Prevents running a filter on `report.rows` which could be undefined.
	const { rows: totalUsersReportRows = [] } = totalUsersReport || {};
	const { rows: trafficSourceReportRows = [] } = trafficSourceReport || {};

	const topTrafficSource =
		trafficSourceReportRows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
			?.dimensionValues?.[ 0 ].value || '-';

	const currentTotalUsers =
		parseInt(
			totalUsersReportRows.filter( makeFilter( 'date_range_0', 0 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value,
			10
		) || 0;
	const currentTopTrafficSourceUsers =
		parseInt(
			trafficSourceReportRows.filter(
				makeFilter( 'date_range_0', 1 )
			)[ 0 ]?.metricValues?.[ 0 ]?.value,
			10
		) || 0;
	const relativeCurrentTopTrafficSourceUsers = currentTotalUsers
		? currentTopTrafficSourceUsers / currentTotalUsers
		: 0;

	const previousTotalUsers =
		parseInt(
			totalUsersReportRows.filter( makeFilter( 'date_range_1', 0 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value,
			10
		) || 0;
	const previousTopTrafficSourceUsers =
		parseInt(
			trafficSourceReportRows.filter(
				makeFilter( 'date_range_1', 1 )
			)[ 0 ]?.metricValues?.[ 0 ]?.value,
			10
		) || 0;
	const relativePreviousTopTrafficSourceUsers = previousTotalUsers
		? previousTopTrafficSourceUsers / previousTotalUsers
		: 0;

	const format = {
		style: 'percent',
		signDisplay: 'never',
		maximumFractionDigits: 1,
	};

	return (
		<MetricTileText
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_TOP_TRAFFIC_SOURCE }
			metricValue={ topTrafficSource }
			metricValueFormat={ format }
			subText={
				// eslint-disable-next-line @wordpress/valid-sprintf
				sprintf(
					/* translators: %d: Percentage of users for the current top traffic source compared to the number of total users for all traffic sources. */
					__( '%s of total traffic', 'google-site-kit' ),
					numFmt( relativeCurrentTopTrafficSourceUsers, format )
				)
			}
			previousValue={ relativePreviousTopTrafficSourceUsers }
			currentValue={ relativeCurrentTopTrafficSourceUsers }
			loading={ loading }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

TopTrafficSourceWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default whenActive( {
	moduleName: 'analytics-4',
	FallbackComponent: ConnectGA4CTATileWidget,
} )( TopTrafficSourceWidget );
