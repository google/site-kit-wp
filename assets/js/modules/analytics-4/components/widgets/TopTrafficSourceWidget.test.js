/**
 * TopTrafficSourceWidget component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	act,
	fireEvent,
	render,
	waitFor,
} from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideKeyMetrics,
	provideModuleRegistrations,
	provideModules,
} from '../../../../../../tests/js/utils';
import {
	getAnalytics4MockResponse,
	provideAnalytics4MockReport,
} from '../../utils/data-mock';
import { getWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import {
	CORE_USER,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import TopTrafficSourceWidget from './TopTrafficSourceWidget';
import WidgetNull from '../../../../googlesitekit/widgets/components/WidgetNull';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { withConnected } from '../../../../googlesitekit/modules/datastore/__fixtures__';

describe( 'TopTrafficSourceWidget', () => {
	const { Widget } = getWidgetComponentProps(
		KM_ANALYTICS_TOP_TRAFFIC_SOURCE
	);

	let registry;
	let dateRangeDates;
	beforeEach( () => {
		registry = createTestRegistry();
		registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );
		dateRangeDates = registry.select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} );
		provideKeyMetrics( registry );
		provideModules( registry, withConnected( 'analytics-4' ) );
	} );

	it( 'renders correctly with the expected metrics', async () => {
		provideAnalytics4MockReport( registry, {
			...dateRangeDates,
			dimensions: [ 'sessionDefaultChannelGroup' ],
			metrics: [
				{
					name: 'totalUsers',
				},
			],
			limit: 1,
			orderBy: 'totalUsers',
		} );
		provideAnalytics4MockReport( registry, {
			...dateRangeDates,
			metrics: [
				{
					name: 'totalUsers',
				},
			],
		} );
		const { container, waitForRegistry } = render(
			<TopTrafficSourceWidget
				Widget={ Widget }
				WidgetNull={ WidgetNull }
			/>,
			{
				registry,
			}
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	it( 'retries both reports when an error is encountered', async () => {
		fetchMock.getOnce(
			new RegExp(
				'^/google-site-kit/v1/modules/analytics-4/data/report.*sessionDefaultChannelGroup.*'
			),
			{
				body: {
					code: 'invalid_json',
					message: 'whatever',
					data: {},
				},
				status: 500,
			}
		);
		fetchMock.getOnce(
			new RegExp(
				'^/google-site-kit/v1/modules/analytics-4/data/report((?!sessionDefaultChannelGroup).)*'
			),
			{
				body: {
					code: 'invalid_json',
					message: 'whatever',
					data: {},
				},
				status: 500,
			}
		);

		provideModuleRegistrations( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );

		const { container, getByText, waitForRegistry } = render(
			<TopTrafficSourceWidget
				Widget={ Widget }
				WidgetNull={ WidgetNull }
			/>,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( console ).toHaveErrored();
		expect( fetchMock ).toHaveFetchedTimes( 2 );
		expect( container ).toMatchSnapshot();

		fetchMock.getOnce(
			new RegExp(
				'^/google-site-kit/v1/modules/analytics-4/data/report.*sessionDefaultChannelGroup.*'
			),
			{
				body: getAnalytics4MockResponse( {
					...dateRangeDates,
					dimensions: [ 'sessionDefaultChannelGroup' ],
					metrics: [
						{
							name: 'totalUsers',
						},
					],
					limit: 1,
					orderBy: 'totalUsers',
				} ),
				status: 200,
			}
		);

		fetchMock.getOnce(
			new RegExp(
				'^/google-site-kit/v1/modules/analytics-4/data/report((?!sessionDefaultChannelGroup).)*'
			),
			{
				body: getAnalytics4MockResponse( {
					...dateRangeDates,
					metrics: [
						{
							name: 'totalUsers',
						},
					],
				} ),
				status: 200,
			}
		);

		await waitForRegistry();

		act( () => {
			fireEvent.click( getByText( 'Retry' ) );
		} );

		await waitFor( () => {
			expect(
				document.querySelector(
					'.googlesitekit-km-widget-tile__metric'
				)
			).toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );
} );
