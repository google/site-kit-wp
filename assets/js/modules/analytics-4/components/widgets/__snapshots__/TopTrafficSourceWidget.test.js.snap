// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopTrafficSourceWidget renders correctly with no data in comparison date range 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsTopTrafficSource googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--text"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-typography googlesitekit-km-widget-tile__title googlesitekit-typography--label googlesitekit-typography--small"
          >
            Top traffic source
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__metric-container"
          >
            <div
              class="googlesitekit-km-widget-tile__metric"
            >
              Organic Search
            </div>
            <p
              class="googlesitekit-km-widget-tile__subtext"
            >
              100% of total traffic
            </p>
          </div>
          <div
            class="googlesitekit-km-widget-tile__metric-change-container"
          >
            <div
              class="googlesitekit-change-badge"
            >
              +100%
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopTrafficSourceWidget should render correctly with the expected metrics 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsTopTrafficSource googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--text"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-typography googlesitekit-km-widget-tile__title googlesitekit-typography--label googlesitekit-typography--small"
          >
            Top traffic source
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__metric-container"
          >
            <div
              class="googlesitekit-km-widget-tile__metric"
            >
              Organic Search
            </div>
            <p
              class="googlesitekit-km-widget-tile__subtext"
            >
              100% of total traffic
            </p>
          </div>
          <div
            class="googlesitekit-km-widget-tile__metric-change-container"
          >
            <div
              class="googlesitekit-change-badge googlesitekit-change-badge--zero"
            >
              0%
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopTrafficSourceWidget should render the generic error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-typography googlesitekit-cta__header_text googlesitekit-typography--label googlesitekit-typography--small"
        >
          Top traffic source
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-typography googlesitekit-cta__title googlesitekit-typography--title googlesitekit-typography--small"
        >
          Data loading failed
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=internal_server_error"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopTrafficSourceWidget should render the insufficient permissions error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-typography googlesitekit-cta__header_text googlesitekit-typography--label googlesitekit-typography--small"
        >
          Top traffic source
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-typography googlesitekit-cta__title googlesitekit-typography--title googlesitekit-typography--small"
        >
          Insufficient permissions
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            Trouble getting access? 
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=analytics-4_insufficient_permissions"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopTrafficSourceWidget should retry both reports when an error is encountered 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-typography googlesitekit-cta__header_text googlesitekit-typography--label googlesitekit-typography--small"
        >
          Top traffic source
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-typography googlesitekit-cta__title googlesitekit-typography--title googlesitekit-typography--small"
        >
          Data loading failed
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <button
            class="mdc-button mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Retry
            </span>
          </button>
          <span
            class="googlesitekit-error-retry-text"
          >
            Retry didnâ€™t work? 
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=invalid_json"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopTrafficSourceWidget should retry both reports when an error is encountered 2`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsTopTrafficSource googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--text"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-typography googlesitekit-km-widget-tile__title googlesitekit-typography--label googlesitekit-typography--small"
          >
            Top traffic source
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Channel (e.g. social, paid, search) that brought in the most visitors to your site"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__metric-container"
          >
            <div
              class="googlesitekit-km-widget-tile__metric"
            >
              Organic Search
            </div>
            <p
              class="googlesitekit-km-widget-tile__subtext"
            >
              100% of total traffic
            </p>
          </div>
          <div
            class="googlesitekit-km-widget-tile__metric-change-container"
          >
            <div
              class="googlesitekit-change-badge googlesitekit-change-badge--zero"
            >
              0%
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
